src/box/applier.cc

```diff
@@ -1252,6 +1298,726 @@ applier_on_rollback(struct trigger *trigger, void *event)
+applier_thread_create(struct applier_thread *thread)
+{
+	static int thread_id = 0;
+	const char *name = tt_sprintf("applier_%d", ++thread_id);
+
+	memset(thread, 0, sizeof(*thread));
+
+	fiber_cond_create(&thread->ready_cond);
+
+	if (cord_costart(&thread->cord, name, applier_thread_f, thread) != 0)
+		diag_raise();
+
+	while (!thread->is_ready) {
+		if (fiber_cond_wait(&thread->ready_cond) != 0)
+			diag_raise();
+	}
+}
+
+/** Alive applier thread list. */
+static RLIST_HEAD(thread_list);
+
+/**
+ * A pointer to the thread which will accept appliers once thread count reaches
+ * the maximum configured value.
+ */
+static struct applier_thread *fill_thread = NULL;
+
+void
+applier_init(void)
+{
+	if (replication_threads == 0)
+		return;
+	for (int i = 0; i < replication_threads; i++) {
+		struct applier_thread *thread =
+			(struct applier_thread *)xmalloc(sizeof(*thread));
+		applier_thread_create(thread);
+		rlist_add_tail_entry(&thread_list, thread, in_list);
+	}
+	fill_thread = rlist_first_entry(&thread_list, struct applier_thread,
+					in_list);
+}
+
+/** Get a working applier thread. */
+static struct applier_thread *
+applier_thread_next(void)
+{
+	struct applier_thread *thread = NULL;
+	assert(!rlist_empty(&thread_list));
+	thread = fill_thread;
+	/*
+	 * Fill the threads in a round-robin manner. Each new applier
+	 * goes to the next available thread.
+	 */
+	if (fill_thread == rlist_last_entry(&thread_list,
+					    struct applier_thread,
+					    in_list)) {
+		fill_thread = rlist_first_entry(&thread_list,
+						struct applier_thread,
+						in_list);
+	} else {
+		fill_thread = rlist_next_entry(fill_thread, in_list);
+	}
+	return thread;
+}
+
+static void
+applier_msg_init(struct applier_msg *msg, struct applier *applier, cmsg_f f)
+{
+	cmsg_init(&msg->base, tx_route);
+	msg->applier = applier;
+	msg->f = f;
+}
+
+/** Initialize applier thread messages. */
+static void
+applier_thread_msgs_init(struct applier *applier)
+{
+	struct applier_thread_data *tdata = applier->tdata;
+	for (int i = 0; i < 2; i++) {
+		struct applier_data_msg *msg = &tdata->msgs[i];
+		memset(msg, 0, sizeof(*msg));
+		applier_msg_init(&msg->base, applier, applier_process_batch);
+		stailq_create(&msg->txs);
+		ibuf_create(&msg->ibuf, &cord()->slabc, 1024);
+		msg->parse_size = 0;
+		region_create(&msg->region, &cord()->slabc);
+	}
+	tdata->msg_ptr = 0;
+
+	applier_msg_init(&tdata->exit_msg.base, applier, applier_exit_thread);
+	diag_create(&tdata->exit_msg.diag);
+}
+
+/**
+ * A trigger fired on each of applier thread fibers death. Propagates the
+ * exception with which the fiber exited to the tx thread.
+ */
+static int
+applier_thread_reader_on_stop(struct trigger *trigger, void *event)
+{
+	struct fiber *fiber = (struct fiber *)event;
+	assert(fiber == fiber());
+	struct applier_thread_data *tdata =
+		(struct applier_thread_data *)trigger->data;
+	assert(fiber == tdata->reader);
+	if (fiber->f_ret != 0 && !(fiber->flags & FIBER_IS_CANCELLED)) {
```

Comment:
Since you need to invoke this code only if the fiber exits normally, I think it'd be more straightforward to execute it directly from the fiber routine rather than use a trigger.